Mathlib/Topology/Germ.lean

```diff
@@ -0,0 +1,190 @@
+    (∀ x, RestrictGermPredicate P A x f) ↔ ∀ᶠ x in 𝓝ˢ A, P x f := by
+  rw [eventually_nhdsSet_iff_forall]
+  rfl
+
+theorem forall_restrictGermPredicate_of_forall
+    {P : ∀ x : X, Germ (𝓝 x) Y → Prop} (h : ∀ x, P x f) :
+    ∀ x, RestrictGermPredicate P A x f :=
+  forall_restrictGermPredicate_iff.mpr (eventually_of_forall h)
+end RestrictGermPredicate
+
+namespace Filter.Germ
+/-- Map the germ at of functions `X × Y → Z` at `p=(x,y) ∈ X × Y` to the corresponding germ
+  of functions `X → Z` at `x ∈ X` -/
+def sliceLeft [TopologicalSpace Y] {p : X × Y} (P : Germ (𝓝 p) Z) : Germ (𝓝 p.1) Z :=
+  P.compTendsto (Prod.mk · p.2) (Continuous.Prod.mk_left p.2).continuousAt
+
+@[simp]
+theorem sliceLeft_coe [TopologicalSpace Y] {y : Y} (f : X × Y → Z) :
+    (↑f : Germ (𝓝 (x, y)) Z).sliceLeft = fun x' ↦ f (x', y) :=
+  rfl
+
+/-- Map the germ at of functions `X × Y → Z` at `p=(x,y) ∈ X × Y` to the corresponding germ
+  of functions `Y → Z` at `y ∈ Y` -/
+def sliceRight [TopologicalSpace Y] {p : X × Y} (P : Germ (𝓝 p) Z) : Germ (𝓝 p.2) Z :=
+  P.compTendsto (Prod.mk p.1) (Continuous.Prod.mk p.1).continuousAt
+
+@[simp]
+theorem sliceRight_coe [TopologicalSpace Y] {y : Y} (f : X × Y → Z) :
+    (↑f : Germ (𝓝 (x, y)) Z).sliceRight = fun y' ↦ f (x, y') :=
+  rfl
+
+/-- The germ of functions `X → Y` at `x ∈ X` is constant w.r.t. the neighbourhood filter `𝓝 x`. -/
+def IsConstant (P : Germ (𝓝 x) Y) : Prop :=
+  P.liftOn (fun f ↦ ∀ᶠ x' in 𝓝 x, f x' = f x) <| by
+    suffices : ∀ f g : X → Y, f =ᶠ[𝓝 x] g → (∀ᶠ x' in 𝓝 x, f x' = f x) → ∀ᶠ x' in 𝓝 x, g x' = g x
+    exact fun f g hfg ↦ propext ⟨fun h ↦ this f g hfg h, fun h ↦ this g f hfg.symm h⟩
+    rintro f g hfg hf
+    refine (hf.and hfg).mono fun x' hx' ↦ ?_
+    rw [← hx'.2, hx'.1, hfg.eq_of_nhds]
+
+theorem isConstant_coe {y} (h : ∀ x', f x' = y) : (↑f : Germ (𝓝 x) Y).IsConstant :=
+  eventually_of_forall fun x' ↦ by rw [h, h]
+
+@[simp]
+theorem isConstant_coe_const {y : Y} : (fun _ : X ↦ y : Germ (𝓝 x) Y).IsConstant :=
+  eventually_of_forall fun _ ↦ rfl
+
+end Filter.Germ
+
+theorem eq_of_germ_isConstant [PreconnectedSpace X]
+    (h : ∀ x : X, (f : Germ (𝓝 x) Y).IsConstant) (x x' : X) : f x = f x' := by
+  revert x
+  erw [← eq_univ_iff_forall]
+  apply IsClopen.eq_univ _ (⟨x', rfl⟩ : {x | f x = f x'}.Nonempty)
+  refine ⟨isOpen_iff_eventually.mpr fun x hx ↦ hx ▸ h x, ?_⟩
+  rw [isClosed_iff_frequently]
+  rintro x hx
+  rcases ((h x).and_frequently hx).exists with ⟨x'', H⟩
+  exact H.1.symm.trans H.2
+
+theorem eq_of_germ_isConstant_on {s : Set X}
+    (h : ∀ x ∈ s, (f : Germ (𝓝 x) Y).IsConstant) (hs : IsPreconnected s) {x' : X} (x_in : x ∈ s)
+    (x'_in : x' ∈ s) : f x = f x' := by
+  haveI := isPreconnected_iff_preconnectedSpace.mp hs
+  let F : s → Y := f ∘ (↑)
+  change F ⟨x, x_in⟩ = F ⟨x', x'_in⟩
+  apply eq_of_germ_isConstant
+  rintro ⟨x, hx⟩
+  have : ContinuousAt ((↑) : s → X) ⟨x, hx⟩ := continuousAt_subtype_val
+  exact this (h x hx)
```

Comment:
These should be proved using [IsLocallyConstant.apply_eq_of_isPreconnected](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/LocallyConstant/Basic.html#IsLocallyConstant.apply_eq_of_isPreconnected), unless this is a too heavy import (which would be a shame)

Reply:
I just tried this: for the first lemma, that's easy. I wouldn't mind a mind with the second one.

Reply:
The second one is taking me further afield that I'd like: I've left a proof_wanted with some hints.